File: androidApp/src/main/java/com/karakasli/kmm_todo_app/android/core/base/BaseViewModel.kt
package com.karakasli.kmm_todo_app.android.core.base

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.karakasli.kmm_todo_app.core.UseCaseResult
import com.karakasli.kmm_todo_app.data.model.error.AppError
import com.karakasli.kmm_todo_app.data.model.event.LoadingEvent
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

/**
 * Created by salihkarakasli on 30.03.2022
 */

open class BaseViewModel : ViewModel() {
    private val _pageLoadingEvent =
        MutableStateFlow<LoadingEvent>(LoadingEvent.LoadingHidden)
    val pageLoadingEvent: StateFlow<LoadingEvent> = _pageLoadingEvent

    protected fun <T> callUseCase(
        callApi: suspend () -> UseCaseResult<T>,
        handleError: (suspend (error: AppError) -> Unit)? = null,
        handleSuccess: (suspend (T?) -> Unit)? = null,
        showLoading: Boolean = false,
        collectDispatcher: CoroutineDispatcher = Dispatchers.Main,
    ) {
        viewModelScope.launch(collectDispatcher) {
            if (showLoading)
                _pageLoadingEvent.value = LoadingEvent.LoadingShown

            when (val result = callApi()) {
                is UseCaseResult.Success -> {
                    handleSuccess?.invoke(result.data)
                }
                is UseCaseResult.ValidationFailed -> {
                    //TODO: Check This part to handle
                    //handleError?.invoke(result.validationError)
                }
                is UseCaseResult.Error -> {
                    handleError?.invoke(result.appError)
                }
            }
            if (showLoading)
                _pageLoadingEvent.value = LoadingEvent.LoadingHidden
        }
    }
}

File: iosApp/iosApp/App/Base/BaseViewModel.swift
//
//  BaseViewModel.swift
//  ToDoTracker
//
//  Created by Salih Karakaşlı on 13.06.2022.
//

import Foundation
import shared

class BaseViewModel: ObservableObject {
    
    @Published var loadingEvent: LoadingEvent? = nil
    @Published var promptEvent: PromptEvent? = nil
    
    func showLoading() {
        loadingEvent = .LoadingShown()
    }
    
    func hideLoading() {
        loadingEvent = .LoadingHidden()
    }
    
    func callUseCase<T>(
        call: () -> UseCaseResult<T>?,
        handleSuccess: ((T) -> Void)?,
        handleError: ((AppError) -> Void)? = nil,
        isShowLoading: Bool = false
    ) {
        if isShowLoading {
            showLoading()
        }
            
        let result = call()
        
        
        if let result = result as? UseCaseResultSuccess {
            if let data = result.data {
                handleSuccess?(data)
            }
        } else if let error = result as? UseCaseResultError {
            handleError?(error.appError)
            print(error.appError.message ?? "error")
        } else if let error = result as? UseCaseResultValidationFailed {
            promptEvent = PromptEvent(title: error.validationError.title, message: error.validationError.message)
        }
        
        if isShowLoading {
            hideLoading()
        }
    }
}
